### PR TITLE
Pardot contact

### PR DESCRIPTION
**Pull request by agentrickard**
## Asana Ticket(s)
- See ticket [Pardot forms in Drupal](https://app.asana.com/0/1174970560648905/1196144381080481/f).

## Description
- Shows one model for displaying Pardot Form Handlers as Drupal blocks.

## Testing instructions

* Checkout this branch
* Inside the VM, run `drush cr` and `drush cim`
* Visit https://global.ddev.site/
* Observe the `Custom Pardot form` block
* Observe the hidden fields 
  - success_location, error_location, full_path, page_title
* Observe the value of `full_path` is set to `Look! A value!` by `/js/bglobal_pardot.js`

## Remaining Steps

* I am not convinced this is the best approach. Using an iFrame for Pardot forms directly still might be better.
* If we use a Form Handler -- https://help.salesforce.com/articleView?id=pardot_form_handlers.htm&type=5 -- then we need to ensure that our forms sync with Pardot fields. We can do that through the theme element introduced here.

In some respected, I repeated part of the FormAPI here. If we need to route the form through Drupal rather than making a direct request to Pardot, we would need to use the FormAPI instead.

- [ ] Determine how we should derive Program Name.
- [ ] Add actual Pardot tracking code.

### CODE REVIEW

File: web/modules/custom/bglobal_pardot/bglobal_pardot.module [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * @file
+ * Drupal hook implementations.
+ */
+
+/**
+ * Implements hook_theme().
+ */
+function bglobal_pardot_theme() {
```
Comment (agentrickard) [inline web/modules/custom/bglobal_pardot/bglobal_pardot.module:11]: This part of the code tells the template what values to expect.

File: web/modules/custom/bglobal_pardot/src/Plugin/Block/PardotFormBlock.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  /**
+   * Build the output.
+   */
+  public function build() {
+    return [
+      '#theme' => 'bglobal_pardot_form',
+      '#action_url' => 'https://go.pardot.com/l/102272/2016-11-04/2klckm',
+      '#success_message' => 'Success!',
+      '#submit_label' => 'Act Now!',
+      '#fields' => [
```
Comment (agentrickard) [inline web/modules/custom/bglobal_pardot/src/Plugin/Block/PardotFormBlock.php:36]: This `#fields` array is what controls the form presentation. We could make this configurable. Or it serves as the framework for variant forms.

File: web/modules/custom/bglobal_pardot/templates/bglobal-pardot-form.html.twig [hunk 3]
```diff
@@ -0,0 +1,58 @@
+{#
```
Comment (agentrickard) [inline web/modules/custom/bglobal_pardot/templates/bglobal-pardot-form.html.twig:1]: This is a module-supplied template which could be over-ridden at the theme layer.

File: web/modules/custom/bglobal_pardot/js/bglobal_pardot.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ */
+function bglobalPardotGetTracker() {
+  hasFired = true;
+  return trackerValue;
+}
+
+/**
+ * Sets the tracker value for the user into local storage.
+ *
+ * @TODO: This code may need to live in another file that is always loaded, or we should
```
Comment (agentrickard) [inline web/modules/custom/bglobal_pardot/js/bglobal_pardot.js:33]: Here's the naive code I setup to run the tracking process.

File: web/modules/custom/bglobal_pardot/bglobal_pardot.module [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  else {
+    /** @var \Drupal\Core\Controller\TitleResolver */
+    $title_resolver = \Drupal::service('title_resolver');
+    /** @var \Symfony\Component\HttpFoundation\Request */
+    $request = \Drupal::request();
+    $page_title = $title_resolver->getTitle($request, $route);
+  }
+
+  $trackers = [
+    'full_path' => 'This value needs to be derived from local storage and set as a script target object.',
```
Comment (agentrickard) [inline web/modules/custom/bglobal_pardot/bglobal_pardot.module:84]: Notice this value is overwritten by the JS.

**Comment by tboggia** [comment]
Just pulled it in, the hidden fields are all filled in as mentioned in the instructions. Will spend some time now looking at the construction.

**Review by tboggia — APPROVED** [review]

**Pull request merged**